tools/autograd/gen_python_functions.py

```diff
@@ -159,90 +260,183 @@ def emit_dispatch(i, function):
+            N = len(outputs)
+            body.append('auto results = r.tensorlist_n<{}>({});'.format(N, arg_idx))
+            for i, arg in enumerate(outputs):
+                formal_args.append('Tensor & {}'.format(arg['name']))
+                actuals.append('results[{}]'.format(i))
+
         env['unpack_args'] = []
         env['formal_args'] = formal_args
-        if unpack_args:
-            unpack_statements_no_default = []
-            unpack_statements_with_default = []
-            actual_names = []
-            for arg, formal_arg, actual in zip(function['arguments'], formal_args, actuals):
-                name = arg['name']
-                actual_names.append(name)
-                unpack_expr = UNPACK_ARG.substitute(formal_arg=formal_arg, actual=actual)
-                if arg.get('python_default_init'):
-                    unpack_statements_with_default.append(unpack_expr)
-                else:
-                    unpack_statements_no_default.append(unpack_expr)
-            env['unpack_args'] = unpack_statements_no_default + unpack_statements_with_default
-            env['actuals'] = actual_names
-            code_template = PY_VARIABLE_CASE_WITH_UNPACK
+        env['actuals'] = actuals
+        if 'call_args' in declaration:
+            env['dispatch_args'] = declaration['call_args']
         else:
-            env['actuals'] = actuals
-            code_template = PY_VARIABLE_CASE
-        if 'call_args' in function:
-            env['dispatch_args'] = function['call_args']
-        else:
-            env['dispatch_args'] = [arg['name'] for arg in function['arguments']]
-        if 'Tensor' in function['method_of']:
+            env['dispatch_args'] = [arg['name'] for arg in declaration['arguments']]
+        if 'Tensor' in declaration['method_of']:
             env['dispatch_args'] = [arg for arg in env['dispatch_args'] if arg != 'self']
-            env['dispatch_call'] = 'self.{}'.format(function['name'])
+            env['dispatch_call'] = 'self.{}'.format(declaration['name'])
         else:
-            env['dispatch_call'] = 'at::{}'.format(function['name'])
+            env['dispatch_call'] = 'at::{}'.format(declaration['name'])
         env['AutoNoGIL'] = 'AutoNoGIL no_gil;'
-        env['AutoGPU'] = auto_gpu(function)
-        env['cond'] = 'if' if i == 0 else '} else if'
-        env = nested_dict(env, function)
+        env['AutoGPU'] = auto_gpu(declaration)
+        env = nested_dict(env, nested_dict(base_env, declaration))
+        body.append(PY_VARIABLE_CALL_DISPATCH.substitute(env))
         py_method_dispatch.append(PY_VARIABLE_DISPATCH.substitute(env))
-        return code_template.substitute(env)
+        return body
 
-    def process_function(name, functions):
+    def emit_dispatch(i, declarations, base_env):
+        if len(declarations) == 1:
+            body = emit_single_dispatch(declarations[0], base_env)
+        else:
+            assert len(declarations) == 2
```

Comment:
nit: the magic length requirements / ordering seems a little hacky to me; can we just use a namedtuple or something to represent this?